loguru/_locks_machinery.py

```diff
@@ -2,13 +2,46 @@
 import threading
 import weakref
 
+
+class HandlerLockNotReentrant(RuntimeError):
+    pass
+
+
+class _HandlerLockWrapper:
```

Comment:
Can we use [`@contextlib.contextmanager`](https://docs.python.org/3/library/contextlib.html#contextlib.contextmanager) for simplification?

Reply:
Yes we can, but only if we separate the Lock from this logic. E.g.

```python
with _acquire_protected(self._lock):
    ...
```

This works given your suggestion to avoid this behaviour on the queued writer.

Another alternative is to keep the wrapper object, and have a contextlib interface like this:

```python
with self._lock.acquire(raise_on_reenter=True):
    ...
```

Although acquire is a poor choice of method name given its meaning on `threading.Lock`, and I can't think of a better name.
```